addon/-private/system/store.js

```diff
@@ -2787,19 +2797,82 @@ function _commit(adapter, store, operation, snapshot) {
   }, label);
 }
 
-function setupRelationships(store, internalModel, data) {
+function inverseRelationshipInitialized(store, internalModel, data, key, modelNameToInverseMap) {
+  let relationshipData = data.relationships[key].data;
+
+  if (!relationshipData) {
+    // can't check inverse for eg { comments: { links: { related: URL }}}
+    return false;
+  }
+
+  let inverseMap = modelNameToInverseMap[internalModel.modelName]
+  if (!inverseMap) {
+    inverseMap = modelNameToInverseMap[internalModel.modelName] = get(internalModel.type, 'inverseMap');
+  }
+  let inverseData = inverseMap[key];
+  if (inverseData === undefined) {
+    inverseData = internalModel.type.inverseFor(key, store);
+  }
+
+  if (!inverseData) {
+    return false;
+  }
+
+  let { name: inverseRelationshipName } = inverseData;
+
+  if (Array.isArray(relationshipData)) {
+    for (let i=0; i<relationshipData.length; ++i) {
+      let inverseInternalModel = store._internalModelsFor(relationshipData[i].type).get(relationshipData[i].id);
+      if (inverseInternalModel && inverseInternalModel._relationships.has(inverseRelationshipName)) {
+        return true;
+      }
+    }
+
+    return false;
+  } else {
+    let inverseInternalModel = store._internalModelsFor(relationshipData.type).get(relationshipData.id);
+    return inverseInternalModel && inverseInternalModel._relationships.has(inverseRelationshipName);
+  }
+}
+
+function setupRelationships(store, internalModel, data, modelNameToInverseMap) {
   if (!data.relationships) {
     return;
   }
 
-  internalModel.type.eachRelationship((key, descriptor) => {
-    if (!data.relationships[key]) {
-      return;
-    }
+  let relationships = internalModel._relationships;
 
-    let relationship = internalModel._relationships.get(key);
-    relationship.push(data.relationships[key]);
-  });
+  for (let key in data.relationships) {
```

Comment:
yep this seems good; also with the setInitial work the backburner.join run in relationships/store/create seems unnecessary